Publisher for multiple sources

caesarxuchao opened the pull request:
Rewrite the publisher to allow gathering contents from multiple sources to the destination repository. For example, `kubernetes/client-go/1.5` gets its content from the `staging/.../1.5` in the **_master**_ branch of the main repository, and `kubernetes/client-go/1.4` gets its content from the `staging/.../1.4` in the **_release-1.4**_ branch of the main repository

Example: https://github.com/caesarxuchao/client-go/commit/a0c3ac3da0ab929aba25d0cdb6cf42885309b06b

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1719)

<!-- Reviewable:end -->

Commit:
rename vendor

Commit:
construct the publishing destination from multiple sources

caesarxuchao commented on mungegithub/publisher/deployment/kubernetes/configmap.yaml:
```diff
@@ -4,13 +4,13 @@ metadata:
   # basic config options.
-  config.organization: kubernetes
+  #config.organization: kubernetes
+  config.organization: caesarxuchao
```
Uncomment this before merging.

caesarxuchao replied:
Done.

mml commented:
I am worried that having logic that crosses the go/shell boundary makes it harder to reason about.  (This is more about `clone.sh` and `construct.sh`, not`publish.sh` )

WDYT of moving the logic of those two into go code?

---

Reviewed 2 of 9 files at r1.
Review status: 2 of 8 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1719#-:-KR_MwFiB1_R3FXs605t:b-6qc3oq)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented:
I don't like the mix either, but some operations are hard in Golang, e.g., [removing an entire directory](http://stackoverflow.com/questions/33450980/golang-remove-all-contents-of-a-directory), or [copy a file](http://stackoverflow.com/questions/21060945/simple-way-to-copy-a-file-in-golang), so I got stuck with those two bash files.

mml commented:
We can still use `exec.Command("rm -rf ...")`, but move them into a go function.  Like

```
func cloneOrDie(destDir, destUrl string) {
...
}
```

---

Review status: 2 of 8 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1719#-:-KR_Wj9MocsKKMDkNxhf:b7yqjzv)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented:
Oh, ok. Do you have other comments? I'll revise when you have done a pass.

mml commented:
Just a few suggestions.

---

Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[mungegithub/mungers/publish.sh, line 55 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1719#-KR_Zi4hsZIy7K-X6Zol:-KR_Zi4inECo-3mq6SIr:bwyobjy) ([raw file](https://github.com/kubernetes/contrib/blob/19584618dffa77ed7c4f5db59c689fce023e24f8/mungegithub/mungers/publish.sh#L55)):*
> ```Shell
>     exit 0
> fi
> git commit -m "published by bot
> ```

It is really awesome to add a link back to the source code here, for some future person trying to figure out "where do these commits come from?"

`published by bot (https://github.com/kubernetes/contrib/tree/master/mungegithub/mungers)`

---

*[mungegithub/mungers/publisher.go, line 32 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1719#-KR__LDZdex-cMJ5aS81:-KR__LDZdex-cMJ5aS82:bwfunqk) ([raw file](https://github.com/kubernetes/contrib/blob/456e375c9165cb29e025039568795f06c8a4e6a4/mungegithub/mungers/publisher.go#L32)):*
> ```Go
> )
> 
> type info struct {
> ```

Should this type be named source?

---

*[mungegithub/mungers/publisher.go, line 42 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1719#-KR___nFpWzIjiZ6KbOQ:-KR___nFpWzIjiZ6KbOR:b-hnl3zm) ([raw file](https://github.com/kubernetes/contrib/blob/456e375c9165cb29e025039568795f06c8a4e6a4/mungegithub/mungers/publisher.go#L42)):*
> ```Go
> 	dstRepo string
> 	// multiple sources mapping to different directories in the dstRepo
> 	srcToDst map[info]string
> ```

Instead of `srcToDst` how about `srcToSubdir`?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1719#-:-KR_adfsvijgSgJ7oUwU:b-lt763)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented:
@mml I've addressed the comments. PTAL. Thank you.

You can see a sample commit here: https://github.com/caesarxuchao/client-go/commit/ebc2f76fc702a0559ed59eedec7168978c8d4a3f

mml commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 9 files at r1, 1 of 3 files at r2, 2 of 4 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1719#-:-KRoOpgswzVbeHE_IUR5:bnfp4nl)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented:
Squashed the commits. I'll deploy the publish after https://github.com/kubernetes/kubernetes/pull/32906 gets merged.

caesarxuchao commented:
Inherit the old lgtm.

foxish commented:
/lgtm
